File: src/interaction/Morse.cpp
/*
  Copyright (C) 2012,2013
      Max Planck Institute for Polymer Research
  Copyright (C) 2008,2009,2010,2011
      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
  
  This file is part of ESPResSo++.
  
  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  
  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>. 
*/

#include "python.hpp"
#include "Morse.hpp"
#include "Tabulated.hpp"
#include "VerletListInteractionTemplate.hpp"
#include "VerletListAdressInteractionTemplate.hpp"
#include "VerletListHadressInteractionTemplate.hpp"
#include "CellListAllPairsInteractionTemplate.hpp"
#include "FixedPairListInteractionTemplate.hpp"

namespace espressopp {
  namespace interaction {
    typedef class VerletListInteractionTemplate< Morse >
    VerletListMorse;
    typedef class VerletListAdressInteractionTemplate< Morse, Tabulated >
    VerletListAdressMorse;
    typedef class VerletListHadressInteractionTemplate< Morse, Tabulated >
    VerletListHadressMorse;
    typedef class CellListAllPairsInteractionTemplate< Morse >
    CellListMorse;
    typedef class FixedPairListInteractionTemplate< Morse >
    FixedPairListMorse;

    //////////////////////////////////////////////////
    // REGISTRATION WITH PYTHON
    //////////////////////////////////////////////////
    void 
    Morse::registerPython() {
      using namespace espressopp::python;

      class_< Morse, bases< Potential > >
    	("interaction_Morse", init< real, real, real, real >())
	.def(init< real, real, real, real, real >())
    	.add_property("epsilon", &Morse::getEpsilon, &Morse::setEpsilon)
    	.add_property("alpha", &Morse::getAlpha, &Morse::setAlpha)
    	.add_property("rMin", &Morse::getRMin, &Morse::setRMin)
    	.def_pickle(Morse_pickle())
    	;

      class_< VerletListMorse, bases< Interaction > > 
        ("interaction_VerletListMorse", init< shared_ptr<VerletList> >())
        .def("setPotential", &VerletListMorse::setPotential)
        .def("getPotential", &VerletListMorse::getPotentialPtr)
      ;

      class_< VerletListAdressMorse, bases< Interaction > >
        ("interaction_VerletListAdressMorse",
                init< shared_ptr<VerletListAdress>, shared_ptr<FixedTupleListAdress> >())
        .def("setPotentialAT", &VerletListAdressMorse::setPotentialAT)
        .def("setPotentialCG", &VerletListAdressMorse::setPotentialCG);
        ;
        
      class_< VerletListHadressMorse, bases< Interaction > >
        ("interaction_VerletListHadressMorse",
                init< shared_ptr<VerletListAdress>, shared_ptr<FixedTupleListAdress> >())
        .def("setPotentialAT", &VerletListHadressMorse::setPotentialAT)
        .def("setPotentialCG", &VerletListHadressMorse::setPotentialCG);
        ;
        
      class_< CellListMorse, bases< Interaction > > 
        ("interaction_CellListMorse", init< shared_ptr< storage::Storage > >())
        .def("setPotential", &CellListMorse::setPotential);
	;

      class_< FixedPairListMorse, bases< Interaction > >
        ("interaction_FixedPairListMorse",
          init< shared_ptr<System>, shared_ptr<FixedPairList>, shared_ptr<Morse> >())
        .def(init< shared_ptr<System>, shared_ptr<FixedPairListAdress>, shared_ptr<Morse> >())
        .def("setPotential", &FixedPairListMorse::setPotential);
        ;
    }
    
  }
}

File: src/interaction/Morse.py
#  Copyright (C) 2012,2013
#      Max Planck Institute for Polymer Research
#  Copyright (C) 2008,2009,2010,2011
#      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
#  
#  This file is part of ESPResSo++.
#  
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>. 


r"""
****************************
espressopp.interaction.Morse
****************************

This class provides methods to compute forces and energies of
the Morse potential.

.. math::

	U = \varepsilon  \left(e^{-2 \alpha (r - r_{min})} - 2 e^{-\alpha (r - r_{min})}\right)






.. function:: espressopp.interaction.Morse(epsilon, alpha, rMin, cutoff, shift)

		:param epsilon: (default: 1.0)
		:param alpha: (default: 1.0)
		:param rMin: (default: 0.0)
		:param cutoff: (default: infinity)
		:param shift: (default: "auto")
		:type epsilon: real
		:type alpha: real
		:type rMin: real
		:type cutoff: 
		:type shift: 

.. function:: espressopp.interaction.VerletListMorse(vl)

		:param vl: 
		:type vl: 

.. function:: espressopp.interaction.VerletListMorse.getPotential(type1, type2)

		:param type1: 
		:param type2: 
		:type type1: 
		:type type2: 
		:rtype: 

.. function:: espressopp.interaction.VerletListMorse.setPotential(type1, type2, potential)

		:param type1: 
		:param type2: 
		:param potential: 
		:type type1: 
		:type type2: 
		:type potential: 

.. function:: espressopp.interaction.VerletListAdressMorse(vl, fixedtupleList)

		:param vl: 
		:param fixedtupleList: 
		:type vl: 
		:type fixedtupleList: 

.. function:: espressopp.interaction.VerletListAdressMorse.setPotentialAT(type1, type2, potential)

		:param type1: 
		:param type2: 
		:param potential: 
		:type type1: 
		:type type2: 
		:type potential: 

.. function:: espressopp.interaction.VerletListAdressMorse.setPotentialCG(type1, type2, potential)

		:param type1: 
		:param type2: 
		:param potential: 
		:type type1: 
		:type type2: 
		:type potential: 

.. function:: espressopp.interaction.VerletListHadressMorse(vl, fixedtupleList)

		:param vl: 
		:param fixedtupleList: 
		:type vl: 
		:type fixedtupleList: 

.. function:: espressopp.interaction.VerletListHadressMorse.setPotentialAT(type1, type2, potential)

		:param type1: 
		:param type2: 
		:param potential: 
		:type type1: 
		:type type2: 
		:type potential: 

.. function:: espressopp.interaction.VerletListHadressMorse.setPotentialCG(type1, type2, potential)

		:param type1: 
		:param type2: 
		:param potential: 
		:type type1: 
		:type type2: 
		:type potential: 

.. function:: espressopp.interaction.CellListMorse(stor)

		:param stor: 
		:type stor: 

.. function:: espressopp.interaction.CellListMorse.setPotential(type1, type2, potential)

		:param type1: 
		:param type2: 
		:param potential: 
		:type type1: 
		:type type2: 
		:type potential: 

.. function:: espressopp.interaction.FixedPairListMorse(system, vl, potential)

		:param system: 
		:param vl: 
		:param potential: 
		:type system: 
		:type vl: 
		:type potential: 

.. function:: espressopp.interaction.FixedPairListMorse.setPotential(potential)

		:param potential: 
		:type potential: 
"""
from espressopp import pmi, infinity
from espressopp.esutil import *

from espressopp.interaction.Potential import *
from espressopp.interaction.Interaction import *
from _espressopp import interaction_Morse, \
                      interaction_VerletListMorse, \
                      interaction_VerletListAdressMorse, \
                      interaction_VerletListHadressMorse, \
                      interaction_CellListMorse, \
                      interaction_FixedPairListMorse

class MorseLocal(PotentialLocal, interaction_Morse):

    def __init__(self, epsilon=1.0, alpha=1.0, rMin=0.0, 
                 cutoff=infinity, shift="auto"):
        """Initialize the local Morse object."""
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            if shift =="auto":
                cxxinit(self, interaction_Morse, 
                        epsilon, alpha, rMin, cutoff)
            else:
                cxxinit(self, interaction_Morse, 
                        epsilon, alpha, rMin, cutoff, shift)

class VerletListMorseLocal(InteractionLocal, interaction_VerletListMorse):

    def __init__(self, vl):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            cxxinit(self, interaction_VerletListMorse, vl)

    def setPotential(self, type1, type2, potential):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            self.cxxclass.setPotential(self, type1, type2, potential)

    def getPotential(self, type1, type2):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            return self.cxxclass.getPotential(self, type1, type2)
        
class VerletListAdressMorseLocal(InteractionLocal, interaction_VerletListAdressMorse):

    def __init__(self, vl, fixedtupleList):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            cxxinit(self, interaction_VerletListAdressMorse, vl, fixedtupleList)

    def setPotentialAT(self, type1, type2, potential):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            self.cxxclass.setPotentialAT(self, type1, type2, potential)

    def setPotentialCG(self, type1, type2, potential):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            self.cxxclass.setPotentialAT(self, type1, type2, potential)
            
class VerletListHadressMorseLocal(InteractionLocal, interaction_VerletListHadressMorse):

    def __init__(self, vl, fixedtupleList):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            cxxinit(self, interaction_VerletListHadressMorse, vl, fixedtupleList)

    def setPotentialAT(self, type1, type2, potential):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            self.cxxclass.setPotentialAT(self, type1, type2, potential)

    def setPotentialCG(self, type1, type2, potential):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            self.cxxclass.setPotentialAT(self, type1, type2, potential)

class CellListMorseLocal(InteractionLocal, interaction_CellListMorse):

    def __init__(self, stor):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            cxxinit(self, interaction_CellListMorse, stor)
        
    def setPotential(self, type1, type2, potential):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            self.cxxclass.setPotential(self, type1, type2, potential)

class FixedPairListMorseLocal(InteractionLocal, interaction_FixedPairListMorse):

    def __init__(self, system, vl, potential):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            cxxinit(self, interaction_FixedPairListMorse, system, vl, potential)
        
    def setPotential(self, potential):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            self.cxxclass.setPotential(self, potential)

if pmi.isController:
    class Morse(Potential):
        'The Morse potential.'
        pmiproxydefs = dict(
            cls = 'espressopp.interaction.MorseLocal',
            pmiproperty = ['epsilon', 'alpha', 'rMin']
            )

    class VerletListMorse(Interaction):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            cls =  'espressopp.interaction.VerletListMorseLocal',
            pmicall = ['setPotential','getPotential']
            )

    class VerletListAdressMorse(Interaction):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            cls =  'espressopp.interaction.VerletListAdressMorseLocal',
            pmicall = ['setPotentialAT', 'setPotentialCG']
            )
            
    class VerletListHadressMorse(Interaction):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            cls =  'espressopp.interaction.VerletListHadressMorseLocal',
            pmicall = ['setPotentialAT', 'setPotentialCG']
            )

    class CellListMorse(Interaction):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            cls =  'espressopp.interaction.CellListMorseLocal',
            pmicall = ['setPotential']
            )

    class FixedPairListMorse(Interaction):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            cls =  'espressopp.interaction.FixedPairListMorseLocal',
            pmicall = ['setPotential']
            )
